src/NumericalAlgorithms/Spectral/Spectral.cpp

```diff
@@ -321,6 +324,118 @@ const Matrix& linear_filter_matrix(const size_t num_points) noexcept {
       LinearFilterMatrixGenerator<BasisType, QuadratureType>>(num_points);
 }
 
+// Synthesize the corresponding functions that take a `Mesh` instead of template
+// parameters.
+
+namespace {
+
+template <typename F>
+decltype(auto) get_spectral_quantity_for_mesh(F&& f,
+                                              const Mesh<1>& mesh) noexcept {
+  const auto num_points = mesh.extents(0);
+  // Switch on runtime values of basis and quadrature to select corresponding
+  // template specialization. We do the switch here because with basis
+  // functions spanning multiple dimensions there could be more logic involved
+  // later.
+  switch (mesh.basis(0)) {
+    case Basis::Legendre:
+      switch (mesh.quadrature(0)) {
+        case Quadrature::Gauss:
+          return f(std::integral_constant<Basis, Basis::Legendre>{},
+                   std::integral_constant<Quadrature, Quadrature::Gauss>{},
+                   num_points);
+          break;
+        case Quadrature::GaussLobatto:
+          return f(
+              std::integral_constant<Basis, Basis::Legendre>{},
+              std::integral_constant<Quadrature, Quadrature::GaussLobatto>{},
+              num_points);
+          break;
+        default:
+          ERROR("Missing quadrature case for spectral quantity");
+      }
+      break;
+    default:
+      ERROR("Missing basis case for spectral quantity");
+  }
+}
+
+}  // namespace
+
```

Comment:
Since so many of these only differ by their function name would it be worth adding two macros, one for `DataVector` and one for `Matrix`? @wthrowe what do you think? I'm not sure it's worth it in this case, but would save a decent amount of duplication.

Reply:
I like the idea of a macro to reduce the code duplication.  I think a `MACRO(name, return type)` could handle all of them except `interpolation_matrix`, which is different enough that it will just have to be written out.

Reply:
Yes, then it'll really reduce code :)